### PR TITLE
bpo-37910: argparse usage wrapping should allow whitespace differences caused by metavar

### PR DESCRIPTION
**Pull request by sjfranklin**
Having metavar be empty (="") can introduce two spaces into the usage summary created by the help text. Here's a minimum working example:

```
import argparse
# based on Vajrasky Kok's script in https://bugs.python.org/issue11874
parser = argparse.ArgumentParser(prog='PROG')
parser.add_argument('--nil', metavar='', required=True)
parser.add_argument('--a', metavar='a' * 165)
parser.parse_args()
```
This would produce the following AssertionError:
```
  File "/minimum_argparse_bug.py", line 7, in <module>
    parser.parse_args()
  File "/path/to/cpython/Lib/argparse.py", line 1758, in parse_args
    args, argv = self.parse_known_args(args, namespace)
  File "/path/to/cpython/Lib/argparse.py", line 1790, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/path/to/cpython/Lib/argparse.py", line 1996, in _parse_known_args
    start_index = consume_optional(start_index)
  File "/path/to/cpython/Lib/argparse.py", line 1936, in consume_optional
    take_action(action, args, option_string)
  File "/path/to/cpython/Lib/argparse.py", line 1864, in take_action
    action(self, namespace, argument_values, option_string)
  File "/path/to/cpython/Lib/argparse.py", line 1037, in __call__
    parser.print_help()
  File "/path/to/cpython/Lib/argparse.py", line 2483, in print_help
    self._print_message(self.format_help(), file)
  File "/path/to/cpython/Lib/argparse.py", line 2467, in format_help
    return formatter.format_help()
  File "/path/to/cpython/Lib/argparse.py", line 281, in format_help
    help = self._root_section.format_help()
  File "/path/to/cpython/Lib/argparse.py", line 212, in format_help
    item_help = join([func(*args) for func, args in self.items])
  File "/path/to/cpython/Lib/argparse.py", line 212, in <listcomp>
    item_help = join([func(*args) for func, args in self.items])
  File "/path/to/cpython/Lib/argparse.py", line 336, in _format_usage
    assert ' '.join(opt_parts) == opt_usage
AssertionError
```
The desired output would be:
```
$ ./python minimum_argparse_bug.py --help
usage: PROG [-h] --nil
            [--a aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa]

optional arguments:
  -h, --help            show this help message and exit
  --nil 
  --a aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```
Changing these two asserts to essentially ignore whitespace fixes the above issue and also allows metavar="\n", ="\t", and a range of other whitespace options.

An more substantial example of special characters now available to metavar:

```
import argparse
# based on Vajrasky Kok's script in https://bugs.python.org/issue11874
parser = argparse.ArgumentParser(prog='PROG')

parser.add_argument('--nil', metavar='', required=True)
parser.add_argument('--Line-Feed', metavar='\n', required=True)
parser.add_argument('--Tab', metavar='\t', required=True)
parser.add_argument('--Carriage-Return', metavar='\r', required=True)
parser.add_argument('--Carriage-Return-and-Line-Feed',
                    metavar='\r\n', required=True)
parser.add_argument('--vLine-Tabulation', metavar='\v', required=True)
parser.add_argument('--x0bLine-Tabulation', metavar='\x0b', required=True)
parser.add_argument('--fForm-Feed', metavar='\f', required=True)
parser.add_argument('--x0cForm-Feed', metavar='\x0c', required=True)
parser.add_argument('--File-Separator', metavar='\x1c', required=True)
parser.add_argument('--Group-Separator', metavar='\x1d', required=True)
parser.add_argument('--Record-Separator', metavar='\x1e', required=True)
parser.add_argument('--C1-Control-Code', metavar='\x85', required=True)
parser.add_argument('--Line-Separator', metavar='\u2028', required=True)
parser.add_argument('--Paragraph-Separator', metavar='\u2029', required=True)
parser.add_argument('--a', metavar='a' * 165)
parser.parse_args()
```
```
./python argparse_bug.py --help
usage: PROG [-h] --nil --Line-Feed --Tab --Carriage-Return --Carriage-Return-and-Line-Feed --vLine-Tabulation
            --x0bLine-Tabulation --fForm-Feed --x0cForm-Feed --File-Separator --Group-Separator
            --Record-Separator --C1-Control-Code --Line-Separator --Paragraph-Separator
            [--a aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa]

optional arguments:
  -h, --help            show this help message and exit
  --nil 
  --Line-Feed 

  --Tab 	
  --Carriage-Return 
  --Carriage-Return-and-Line-Feed 

  --vLine-Tabulation 
  --x0bLine-Tabulation 
  --fForm-Feed 
  --x0cForm-Feed 
  --File-Separator 
  --Group-Separator 
  --Record-Separator 
  --C1-Control-Code 
  --Line-Separator  
  --Paragraph-Separator  
  --a aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```

I tested this fix on Linux (RHEL 7.6; 3.10.0-957) with yesterday's master branch (Python 3.9.0a0). The fix was also compatible when I manually changed argparse.py on Python 3.7.3 installed through Miniconda3.

https://bugs.python.org/issue37910

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Our records indicate we have not received your CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

File: Misc/NEWS.d/next/Library/2019-08-21-23-10-12.bpo-37910.dByB09.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+The two asserts in _format_usage now ignore extra whitespaces which are used to split a long usages into multiple lines.
```
Comment (epicfaace) [inline Misc/NEWS.d/next/Library/2019-08-21-23-10-12.bpo-37910.dByB09.rst]: Can you add more context here so it's clear that this affects argparse, and what behavior it allows argparse to do?

File: Lib/argparse.py [hunk 2]
```diff
@@ -333,8 +333,11 @@ def _format_usage(self, usage, actions, groups, prefix):
                 pos_usage = format(positionals, groups)
                 opt_parts = _re.findall(part_regexp, opt_usage)
                 pos_parts = _re.findall(part_regexp, pos_usage)
-                assert ' '.join(opt_parts) == opt_usage
-                assert ' '.join(pos_parts) == pos_usage
+
+                # ignore extra whitespace differences
```
Comment (epicfaace) [inline Lib/argparse.py:339]: Can you add some tests in test_argparse.py verifying this behavior?

**Review by epicfaace — CHANGES_REQUESTED** [review]

File: Misc/NEWS.d/next/Library/2019-08-21-23-10-12.bpo-37910.dByB09.rst (hunk 1 above)
Comment (sjfranklin) [reply Misc/NEWS.d/next/Library/2019-08-21-23-10-12.bpo-37910.dByB09.rst]: Sure, I'll try to do this soon. Thanks for noticing that!

File: Lib/argparse.py (hunk 2 above)
Comment (sjfranklin) [reply Lib/argparse.py:339]: Absolutely. It might not be too soon since work's a bit busy at the moment.

Comment (sjfranklin) [reply Lib/argparse.py:339]: Thanks for your review, @epicfaace. I added several new tests for a range of special characters.

**Review by epicfaace — APPROVED** [review]

File: Lib/test/test_argparse.py [hunk 3]
```diff
@@ -5224,6 +5224,192 @@ def test_nargs_3_metavar_length3(self):
         self.do_test_no_exception(nargs=3, metavar=("1", "2", "3"))
 
 
+class TestAddArgumentMetavarWrapNoException(TestCase):
+    """Check that certain special character wrap with no exceptions
+    Based off TestAddArgumentMetavar"""
+
```
Comment (rhettinger) [inline Lib/test/test_argparse.py:5230]: Please combine all of these into a single test method that runs the cases in a loop with unittest subtests.

File: Misc/NEWS.d/next/Library/2019-09-20-01-08-53.bpo-37910.hxlo0G.rst [hunk 4]
```diff
@@ -0,0 +1,3 @@
+argparse.py now allows metavar to be certain whitespace characters, such as
```
Comment (shihai1991) [inline Misc/NEWS.d/next/Library/2019-09-20-01-08-53.bpo-37910.hxlo0G.rst:1]: pls use `:mod:argparse` in here.

**Review by shihai1991 — COMMENTED** [review]
Thanks,  It works.

**Comment by hauntsaninja** [comment]
@sjfranklin are you interested in seeing this PR through?

**Comment by erlend-aasland** [comment]
Closing this, as the linked issue was resolved by #105039. Thanks for the PR, though!

**Pull request closed without merging**